File: BloomreachSDKMauiAndroid/lib/src/test/java/com/bloomreach/sdk/maui/android/NotificationsApiTests.kt
package com.bloomreach.sdk.maui.android

import android.content.Intent
import com.exponea.sdk.Exponea
import com.exponea.sdk.models.NotificationAction
import com.exponea.sdk.models.NotificationData
import io.mockk.every
import io.mockk.slot
import io.mockk.verify
import org.junit.Assert
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class NotificationsApiTests : TestsBase() {

    @Test
    fun HandleRemoteMessage() {
        val payloadSlot = slot<Map<String, String>>()
        val result = instance.invokeMethod(
            "HandleRemoteMessage",
            readTestFile("SamplePushNotification")
        )
        verify {
            Exponea.handleRemoteMessage(
                any(),
                capture(payloadSlot),
                any(),
                true
            )
        }
        assertTrue(result.success)
        assertTrue(payloadSlot.isCaptured)
        Assert.assertNotNull(payloadSlot.captured)
        val capturedMap = payloadSlot.captured
        Assert.assertNotNull(capturedMap["url_ios"])
        Assert.assertNotNull(capturedMap["legacy_ios_category"])
        Assert.assertNotNull(capturedMap["title"])
        Assert.assertNotNull(capturedMap["action"])
        Assert.assertNotNull(capturedMap["message"])
        Assert.assertNotNull(capturedMap["image"])
        Assert.assertNotNull(capturedMap["actions"])
        Assert.assertNotNull(capturedMap["sound"])
        Assert.assertNotNull(capturedMap["aps"])
        Assert.assertNotNull(capturedMap["url_params"])
        Assert.assertNotNull(capturedMap["source"])
        Assert.assertNotNull(capturedMap["silent"])
        Assert.assertNotNull(capturedMap["has_tracking_consent"])
        Assert.assertNotNull(capturedMap["consent_category_tracking"])
    }

    @Test
    fun HandlePushNotificationOpened() {
        val dataSlot = slot<NotificationData>()
        val actionSlot = slot<NotificationAction>()
        val result = instance.invokeMethod(
            "HandlePushNotificationOpened",
            readTestFile("HandlePushNotificationOpened")
        )
        verify {
            Exponea.trackClickedPush(
                capture(dataSlot),
                capture(actionSlot),
                any()
            )
        }
        assertTrue(result.success)
        assertTrue(dataSlot.isCaptured)
        assertTrue(actionSlot.isCaptured)
        assertEquals("val1", dataSlot.captured.attributes["prop1"])
        assertEquals("override", dataSlot.captured.attributes["prop2"])
        assertEquals("button", actionSlot.captured.actionType)
        assertEquals("Click me", actionSlot.captured.actionName)
        assertEquals("https://google.com", actionSlot.captured.url)
    }

    @Test
    fun HandlePushNotificationOpenedWithoutTrackingConsent() {
        val dataSlot = slot<NotificationData>()
        val actionSlot = slot<NotificationAction>()
        val result = instance.invokeMethod(
            "HandlePushNotificationOpenedWithoutTrackingConsent",
            readTestFile("HandlePushNotificationOpenedWithoutTrackingConsent")
        )
        verify {
            Exponea.trackClickedPushWithoutTrackingConsent(
                capture(dataSlot),
                capture(actionSlot),
                any()
            )
        }
        assertTrue(result.success)
        assertTrue(dataSlot.isCaptured)
        assertTrue(actionSlot.isCaptured)
        assertEquals("val1", dataSlot.captured.attributes["prop1"])
        assertEquals("override", dataSlot.captured.attributes["prop2"])
        assertEquals("button", actionSlot.captured.actionType)
        assertEquals("Click me", actionSlot.captured.actionName)
        assertEquals("https://google.com", actionSlot.captured.url)
    }

    @Test
    fun HandleCampaignClick() {
        val intentSlot = slot<Intent>()
        val result = instance.invokeMethod(
            "HandleCampaignClick",
            readTestFile("HandleCampaignClick")
        )
        verify {
            Exponea.handleCampaignIntent(
                capture(intentSlot),
                any()
            )
        }
        assertTrue(result.success)
        assertEquals(Intent.ACTION_VIEW, intentSlot.captured.action)
        assertEquals("https://google.com/unit/test/with?params=1&orparam=true", intentSlot.captured.data?.toString())
    }

    @Test
    fun HandleHmsPushToken() {
        val result = instance.invokeMethod(
            "HandleHmsPushToken",
            readTestFile("HandleHmsPushToken")
        )
        verify {
            Exponea.handleNewHmsToken(
                any(),
                "hms-token-12345"
            )
        }
        assertTrue(result.success)
    }

    @Test
    fun HandlePushToken() {
        val result = instance.invokeMethod(
            "HandlePushToken",
            readTestFile("HandlePushToken")
        )
        verify {
            Exponea.handleNewToken(
                any(),
                "gms-token-12345"
            )
        }
        assertTrue(result.success)
    }

    @Test
    fun IsBloomreachNotification_Valid() {
        every {
            Exponea.isExponeaPushNotification(any())
        } returns true
        val result = instance.invokeMethod(
            "IsBloomreachNotification",
            readTestFile("IsBloomreachNotification")
        )
        verify {
            Exponea.isExponeaPushNotification(
                mapOf(
                    "prop1" to "val1",
                    "prop2" to "val2"
                )
            )
        }
        assertTrue(result.success)
        assertEquals("true", result.data)
    }

    @Test
    fun IsBloomreachNotification_Invalid() {
        every {
            Exponea.isExponeaPushNotification(any())
        } returns false
        val result = instance.invokeMethod(
            "IsBloomreachNotification",
            readTestFile("IsBloomreachNotification")
        )
        verify {
            Exponea.isExponeaPushNotification(
                mapOf(
                    "prop1" to "val1",
                    "prop2" to "val2"
                )
            )
        }
        assertTrue(result.success)
        assertEquals("false", result.data)
    }

    @Test
    fun TrackClickedPush() {
        val dataSlot = slot<NotificationData>()
        val actionSlot = slot<NotificationAction>()
        val result = instance.invokeMethod(
            "TrackClickedPush",
            readTestFile("TrackClickedPush")
        )
        verify {
            Exponea.trackClickedPush(
                capture(dataSlot),
                capture(actionSlot),
                any()
            )
        }
        assertTrue(result.success)
        assertTrue(dataSlot.isCaptured)
        assertTrue(actionSlot.isCaptured)
        assertEquals("val1", dataSlot.captured.attributes["prop1"])
        assertEquals("override", dataSlot.captured.attributes["prop2"])
        assertEquals("button", actionSlot.captured.actionType)
        assertEquals("Click me", actionSlot.captured.actionName)
        assertEquals("https://google.com", actionSlot.captured.url)
    }

    @Test
    fun TrackClickedPushWithoutTrackingConsent() {
        val dataSlot = slot<NotificationData>()
        val actionSlot = slot<NotificationAction>()
        val result = instance.invokeMethod(
            "TrackClickedPushWithoutTrackingConsent",
            readTestFile("TrackClickedPushWithoutTrackingConsent")
        )
        verify {
            Exponea.trackClickedPushWithoutTrackingConsent(
                capture(dataSlot),
                capture(actionSlot),
                any()
            )
        }
        assertTrue(result.success)
        assertTrue(dataSlot.isCaptured)
        assertTrue(actionSlot.isCaptured)
        assertEquals("val1", dataSlot.captured.attributes["prop1"])
        assertEquals("override", dataSlot.captured.attributes["prop2"])
        assertEquals("button", actionSlot.captured.actionType)
        assertEquals("Click me", actionSlot.captured.actionName)
        assertEquals("https://google.com", actionSlot.captured.url)
    }

    @Test
    fun TrackPushToken() {
        val result = instance.invokeMethod(
            "TrackPushToken",
            readTestFile("HandlePushToken")
        )
        verify {
            Exponea.trackPushToken(
                "gms-token-12345"
            )
        }
        assertTrue(result.success)
    }

    @Test
    fun TrackHmsPushToken() {
        val result = instance.invokeMethod(
            "TrackHmsPushToken",
            readTestFile("HandleHmsPushToken")
        )
        verify {
            Exponea.trackHmsPushToken(
                "hms-token-12345"
            )
        }
        assertTrue(result.success)
    }

    @Test
    fun TrackDeliveredPush() {
        val dataSlot = slot<NotificationData>()
        val result = instance.invokeMethod(
            "TrackDeliveredPush",
            readTestFile("TrackDeliveredPush")
        )
        verify {
            Exponea.trackDeliveredPush(
                capture(dataSlot),
                any()
            )
        }
        assertTrue(result.success)
        assertTrue(dataSlot.isCaptured)
        Assert.assertNotNull(dataSlot.captured)
        assertEquals("val1", dataSlot.captured.attributes["prop1"])
        assertEquals("val2", dataSlot.captured.attributes["prop2"])
    }

    @Test
    fun TrackDeliveredPushWithoutTrackingConsent() {
        val dataSlot = slot<NotificationData>()
        val result = instance.invokeMethod(
            "TrackDeliveredPushWithoutTrackingConsent",
            readTestFile("TrackDeliveredPush")
        )
        verify {
            Exponea.trackDeliveredPushWithoutTrackingConsent(
                capture(dataSlot),
                any()
            )
        }
        assertTrue(result.success)
        assertTrue(dataSlot.isCaptured)
        Assert.assertNotNull(dataSlot.captured)
        assertEquals("val1", dataSlot.captured.attributes["prop1"])
        assertEquals("val2", dataSlot.captured.attributes["prop2"])
    }
}

File: BloomreachTests/NotificationsApiTests.cs
using BloomreachTests.Utils;
using Bloomreach;
using Bloomreach.Utils;

namespace BloomreachTests;

public class NotificationsApiTests
{

    private readonly MethodInvokeCollector _methodCollector = new MethodInvokeCollector();
    
    [SetUp]
    public void Setup()
    {
        _methodCollector.RegisterDefaultMethodResult(new MethodMauiResult(
            success: true,
            data: "",
            error: ""
        ));
        BloomreachSDK.Instance = new BloomreachSdkMock(_methodCollector);
    }
    [TearDown]
    public void TearDown()
    {
        _methodCollector.Clear();
    }

    [Test]
    public void HandlePushNotificationOpened()
    {
        var notificationAction = new NotificationAction(
                "button",
                "Click me",
                "https://google.com"
            )
            .WithAttribute("prop1", "val1")
            .WithAttribute("prop2", 2)
            .WithAttribute("prop3", null)
            .WithAttribute("prop2", "override");
        BloomreachSDK.HandlePushNotificationOpened(notificationAction);
        _methodCollector.VerifyMethodCalled("HandlePushNotificationOpened");
        _methodCollector.VerifyMethodInput(
            "HandlePushNotificationOpened",
            TestUtils.ReadFile("HandlePushNotificationOpened")
        );
    }
    
    [Test]
    public void HandlePushNotificationOpenedWithoutTrackingConsent()
    {
        var notificationAction = new NotificationAction(
                "button",
                "Click me",
                "https://google.com"
            )
            .WithAttribute("prop1", "val1")
            .WithAttribute("prop2", 2)
            .WithAttribute("prop3", null)
            .WithAttribute("prop2", "override");
        BloomreachSDK.HandlePushNotificationOpenedWithoutTrackingConsent(notificationAction);
        _methodCollector.VerifyMethodCalled("HandlePushNotificationOpenedWithoutTrackingConsent");
        _methodCollector.VerifyMethodInput(
            "HandlePushNotificationOpenedWithoutTrackingConsent",
            TestUtils.ReadFile("HandlePushNotificationOpenedWithoutTrackingConsent")
        );
    }
    
    [Test]
    public void HandleCampaignClick()
    {
        var campaignUrl = new Uri("https://google.com/unit/test/with?params=1&orparam=true");
        BloomreachSDK.HandleCampaignClick(campaignUrl);
        _methodCollector.VerifyMethodCalled("HandleCampaignClick");
        _methodCollector.VerifyMethodInput(
            "HandleCampaignClick",
            TestUtils.ReadFile("HandleCampaignClick")
        );
    }
    
    [Test]
    public void HandleHmsPushToken()
    {
        BloomreachSDK.HandleHmsPushToken("hms-token-12345");
        _methodCollector.VerifyMethodCalled("HandleHmsPushToken");
        _methodCollector.VerifyMethodInput(
            "HandleHmsPushToken",
            TestUtils.ReadFile("HandleHmsPushToken")
        );
    }
    
    [Test]
    public void HandlePushToken()
    {
        BloomreachSDK.HandlePushToken("gms-token-12345");
        _methodCollector.VerifyMethodCalled("HandlePushToken");
        _methodCollector.VerifyMethodInput(
            "HandlePushToken",
            TestUtils.ReadFile("HandlePushToken")
        );
    }
    
    [Test]
    public void IsBloomreachNotification_TRUE()
    {
        _methodCollector.RegisterSuccessMethodResult("IsBloomreachNotification", "true");
        NotificationPayload notificationPayload = NotificationPayload.Parse(new Dictionary<string, string>()
        {
            {"prop1", "val1"},
            {"prop2", "val2"}
        });
        var result = BloomreachSDK.IsBloomreachNotification(notificationPayload);
        _methodCollector.VerifyMethodCalled("IsBloomreachNotification");
        _methodCollector.VerifyMethodInput(
            "IsBloomreachNotification",
            TestUtils.ReadFile("IsBloomreachNotification")
        );
        Assert.That(result, Is.True);
    }
    
    [Test]
    public void IsBloomreachNotification_FALSE()
    {
        _methodCollector.RegisterSuccessMethodResult("IsBloomreachNotification", "false");
        NotificationPayload notificationPayload = NotificationPayload.Parse(new Dictionary<string, string>()
        {
            {"prop1", "val1"},
            {"prop2", "val2"}
        });
        var result = BloomreachSDK.IsBloomreachNotification(notificationPayload);
        _methodCollector.VerifyMethodCalled("IsBloomreachNotification");
        _methodCollector.VerifyMethodInput(
            "IsBloomreachNotification",
            TestUtils.ReadFile("IsBloomreachNotification")
        );
        Assert.That(result, Is.False);
    }
    
    [Test]
    public void IsBloomreachNotification_InvalidIsFalse()
    {
        _methodCollector.RegisterSuccessMethodResult("IsBloomreachNotification", "non-valid-value");
        NotificationPayload notificationPayload = NotificationPayload.Parse(new Dictionary<string, string>()
        {
            {"prop1", "val1"},
            {"prop2", "val2"}
        });
        var result = BloomreachSDK.IsBloomreachNotification(notificationPayload);
        _methodCollector.VerifyMethodCalled("IsBloomreachNotification");
        _methodCollector.VerifyMethodInput(
            "IsBloomreachNotification",
            TestUtils.ReadFile("IsBloomreachNotification")
        );
        Assert.That(result, Is.False);
    }
    
    [Test]
    public void IsBloomreachNotification_ErrorIsFalse()
    {
        _methodCollector.RegisterFailureMethodResult("IsBloomreachNotification", "some error");
        NotificationPayload notificationPayload = NotificationPayload.Parse(new Dictionary<string, string>()
        {
            {"prop1", "val1"},
            {"prop2", "val2"}
        });
        var result = BloomreachSDK.IsBloomreachNotification(notificationPayload);
        _methodCollector.VerifyMethodCalled("IsBloomreachNotification");
        _methodCollector.VerifyMethodInput(
            "IsBloomreachNotification",
            TestUtils.ReadFile("IsBloomreachNotification")
        );
        Assert.That(result, Is.False);
    }
    
    [Test]
    public void TrackClickedPush()
    {
        var notificationAction = new NotificationAction(
                "button",
                "Click me",
                "https://google.com"
            )
            .WithAttribute("prop1", "val1")
            .WithAttribute("prop2", 2)
            .WithAttribute("prop3", null)
            .WithAttribute("prop2", "override");
        BloomreachSDK.TrackClickedPush(notificationAction);
        _methodCollector.VerifyMethodCalled("TrackClickedPush");
        _methodCollector.VerifyMethodInput(
            "TrackClickedPush",
            TestUtils.ReadFile("TrackClickedPush")
        );
    }
    
    [Test]
    public void TrackClickedPushWithoutTrackingConsent()
    {
        var notificationAction = new NotificationAction(
                "button",
                "Click me",
                "https://google.com"
            )
            .WithAttribute("prop1", "val1")
            .WithAttribute("prop2", 2)
            .WithAttribute("prop3", null)
            .WithAttribute("prop2", "override");
        BloomreachSDK.TrackClickedPushWithoutTrackingConsent(notificationAction);
        _methodCollector.VerifyMethodCalled("TrackClickedPushWithoutTrackingConsent");
        _methodCollector.VerifyMethodInput(
            "TrackClickedPushWithoutTrackingConsent",
            TestUtils.ReadFile("TrackClickedPushWithoutTrackingConsent")
        );
    }
    
    [Test]
    public void TrackPushToken()
    {
        BloomreachSDK.TrackPushToken("gms-token-12345");
        _methodCollector.VerifyMethodCalled("TrackPushToken");
        _methodCollector.VerifyMethodInput(
            "TrackPushToken",
            TestUtils.ReadFile("HandlePushToken")
        );
    }
    
    [Test]
    public void TrackHmsPushToken()
    {
        BloomreachSDK.TrackHmsPushToken("hms-token-12345");
        _methodCollector.VerifyMethodCalled("TrackHmsPushToken");
        _methodCollector.VerifyMethodInput(
            "TrackHmsPushToken",
            TestUtils.ReadFile("HandleHmsPushToken")
        );
    }

    [Test]
    public void TrackDeliveredPush()
    {
        NotificationPayload notificationPayload = NotificationPayload.Parse(new Dictionary<string, string>()
        {
            {"attributes", ConverterUtils.SerializeInput(new Dictionary<string, string>
            {
                {"prop1", "val1"},
                {"prop2", "val2"}
            })!}
        });
        BloomreachSDK.TrackDeliveredPush(notificationPayload);
        _methodCollector.VerifyMethodCalled("TrackDeliveredPush");
        _methodCollector.VerifyMethodInput(
            "TrackDeliveredPush",
            TestUtils.ReadFile("TrackDeliveredPush")
        );
    }
    
    [Test]
    public void TrackDeliveredPushWithoutTrackingConsent()
    {
        NotificationPayload notificationPayload = NotificationPayload.Parse(new Dictionary<string, string>()
        {
            {"attributes", ConverterUtils.SerializeInput(new Dictionary<string, string>
            {
                {"prop1", "val1"},
                {"prop2", "val2"}
            })!}
        });
        BloomreachSDK.TrackDeliveredPushWithoutTrackingConsent(notificationPayload);
        _methodCollector.VerifyMethodCalled("TrackDeliveredPushWithoutTrackingConsent");
        _methodCollector.VerifyMethodInput(
            "TrackDeliveredPushWithoutTrackingConsent",
            TestUtils.ReadFile("TrackDeliveredPush")
        );
    }
    
    [Test]
    public void RequestPushAuthorization()
    {
        BloomreachSDK.RequestPushAuthorization();
        _methodCollector.VerifyMethodCalled("RequestPushAuthorization");
    }

    [Test]
    public void SetReceivedPushCallback()
    {
        _methodCollector.RegisterSuccessMethodResult("SetReceivedPushCallback", TestUtils.ReadFile("SetReceivedPushCallback"));
        NotificationPayload? result = null;
        BloomreachSDK.SetReceivedPushCallback(payload =>
        {
            result = payload;
        });
        _methodCollector.VerifyMethodCalled("SetReceivedPushCallback");
        Thread.Sleep(TimeSpan.FromSeconds(1));
        Assert.That(
            result,
            Is.Not.Null
        );
        Assert.That(
            result!.RawData,
            Is.EqualTo(new Dictionary<string, string>
            {
                {"prop1", "val1"},
                {"prop2", "val2"}
            })
        );
    }

    [Test]
    public void SetOpenedPushCallback()
    {
        _methodCollector.RegisterSuccessMethodResult("SetOpenedPushCallback", TestUtils.ReadFile("SetOpenedPushCallback"));
        NotificationAction? result = null;
        BloomreachSDK.SetOpenedPushCallback(action =>
        {
            result = action;
        });
        _methodCollector.VerifyMethodCalled("SetOpenedPushCallback");
        Thread.Sleep(TimeSpan.FromSeconds(1));
        Assert.That(
            result,
            Is.Not.Null
        );
        Assert.That(result!.ActionType, Is.EqualTo("button"));
        Assert.That(result!.ActionName, Is.EqualTo("Click me"));
        Assert.That(result!.Url, Is.EqualTo("https://google.com"));
        Assert.That(
            result!.Attributes,
            Is.EqualTo(new Dictionary<string, object?>
            {
                {"prop1", "val1"},
                {"prop2", "override"}
            })
        );
    }

}
